File: tests/ops/test_max_pool2d.cpp
#include "convolution/max_pool_2d_cpu.h"
#include <iostream>
#include <memory>
int main() {
    MaxPool2dCpu ce(2);

    auto input = std::make_shared<Tensor>(std::vector<int>{1, 1, 4, 4});

    for (int i = 0; i < 16; i++) {
        input->data[i] = i + 1;
    }

    std::cout << "input: " << input << std::endl;

    auto output = ce.Forward(input);
    std::cout << "output: " << output << std::endl;

    auto output_grad = std::make_shared<Tensor>(std::vector<int>{1, 1, 2, 2});
    for (int i = 0; i < 4; i++) {
        output_grad->data[i] = i + 1;
    }

    auto grad = ce.Backward(output_grad);
    std::cout << "Gradient: " << grad << std::endl;
    return 0;
}

File: python/test_maxpool2d.py
import torch.nn as nn
import torch

mp = nn.MaxPool2d(2, stride=2)

input = torch.tensor([[[[1, 2, 3, 4],
                            [5, 6, 7, 8],
                            [9, 10, 11, 12],
                            [13, 14, 15, 16]]]],dtype=torch.float32,requires_grad=True)

output = mp(input)
print(input)
print(output)

a = torch.tensor([[[[1, 2],
                            [3, 4]]]],dtype=torch.float32,requires_grad=True)
output.backward(a)
print(input.grad)
